cpp/src/text/generate_ngrams.cu

```diff
@@ -0,0 +1,162 @@
+        char* out_ptr = d_chars ? d_chars + d_offsets[idx] : nullptr;
+        cudf::size_type bytes = 0;
+        for( cudf::size_type n=0; n < ngrams; ++n )
+        {
+            auto const d_str = d_strings.element<cudf::string_view>(n+idx);
+            bytes += d_str.size_bytes();
+            if( out_ptr )
+                out_ptr = cudf::strings::detail::copy_string(out_ptr,d_str);
+            if( (n+1) >= ngrams )
+                continue;
+            bytes += d_separator.size_bytes();
+            if( out_ptr )
+                out_ptr = cudf::strings::detail::copy_string(out_ptr,d_separator);
+        }
+        return bytes;
+    }
+};
+
+} // namespace
+
+std::unique_ptr<cudf::column> generate_ngrams( cudf::strings_column_view const& strings,
+                                               cudf::size_type ngrams = 2,
+                                               cudf::string_scalar const& separator = cudf::string_scalar{"_"},
+                                               rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                                               cudaStream_t stream = 0 )
+{
+    CUDF_EXPECTS( separator.is_valid(), "Parameter separator must be valid");
+    cudf::string_view d_separator( separator.data(), separator.size() );
+    CUDF_EXPECTS( ngrams > 1, "Parameter ngrams should be an integer value of 2 or greater");
+
+    auto strings_count = strings.size();
+    if( strings_count==0 ) // if no strings, return an empty column
+        return cudf::make_empty_column(cudf::data_type{cudf::STRING});
+
+    auto execpol = rmm::exec_policy(stream);
+    auto strings_column = cudf::column_device_view::create(strings.parent(),stream);
+    auto d_strings = *strings_column;
+
+    // first create a new offsets vector removing nulls and empty strings from the input column
+    std::unique_ptr<cudf::column> non_empty_offsets_column = [&] {
+        cudf::column_view offsets_view( cudf::data_type{cudf::INT32}, strings_count+1, 
+                                        strings.offsets().data<int32_t>() );
+        auto table_offsets = cudf::experimental::detail::copy_if( cudf::table_view({offsets_view}),
+                [d_strings, strings_count] __device__ (cudf::size_type idx) {
+                    if( idx==strings_count )
+                        return true;
+                    if( d_strings.is_null(idx) )
+                        return false;
+                    return !d_strings.element<cudf::string_view>(idx).empty();
+                }, mr, stream )->release();
+        strings_count = table_offsets.front()->size()-1;
+        return std::move(table_offsets.front());
+    } (); // this allows freeint the temporary table_offsets
+
+    if( strings_count==0 ) // if no non-empty strings, return an empty column
+        return cudf::make_empty_column(cudf::data_type{cudf::STRING});
+    CUDF_EXPECTS( strings_count >= ngrams , "Insufficient number of strings to generate ngrams");
```

Comment:
Should strings_count == 0 and (for example) strings_count==1 be treated differently? AFAICT they both represent the same case - insufficient number of elements. 
Either strings_count < ngrams should yield an empty column, or empty input should cause an exception.

Reply:
Apparently this is a common scenario when using Dask and we were told not to throw exceptions for empty columns.

Reply:
> Apparently this is a common scenario when using Dask and we were told not to throw exceptions for empty columns.

What about the other option, returning an empty column when there aren't enough elements to form a single ngram?

Reply:
I think your first statement makes more sense. Here an empty column was indeed not passed but only _became empty_ by the filter and so therefore we should be allowed to throw an exception as per the ngram count check. This is cleaner and does not violate the empty-column rule.